cluster-autoscaler/proposals/buffers.md

```diff
@@ -0,0 +1,576 @@
+    // In the future the cloud providers will be able to offer their own buffering strategies.
+    ProvisioningStrategy string
+}
+
+type NodeAutoscalingBufferCapacity struct {
+    // One of the following must be set
+    // +optional
+    Replicas *ReplicaBufferCapacity
+    // +optional
+    Resources *ResourceBufferCapacity
+    // +optional
+    Pods *PodBufferCapacity
+}
+
+type ReplicaBufferCapacity struct {
+	// An object implementing scale subresource (for example deployment, job etc.)
+    TargetRef *TypedLocalObjectReference
+    // One of the following must be set:
+    // +optional
+    Exactly *int
+    // +optional
+    Percent *PercentReplicaBufferCapacity
+}
+
+type PercentReplicaBufferCapacity struct {
+    Percent  int
+    MinCount *int
+    MaxCount *int
+}
+
+type ResourceBufferCapacity struct {
+	// Defines where the resource should be created.
+	TargetRef *TypedLocalObjectReference
+    // Upper limit of the total capacity that will be scaled for this buffer.
+    // The total capacity will be split into smaller chunks that each fits onto nodes.
+    // The size of a chunk will be either decided for the user (using some defaults and heuristics) or the
+    // user can specify per pod Requests to define the shape of the buffer space. In the latter case the
+    // CPU and Mem will be the maximum size of the buffer and if the Requests are too big to fit the total
+    // buffer size the admission of the buffer should be denied.
+    TotalCpu resource.Quantity
+    TotalMem resource.Quantity
+
+    // Optional if defining a single chunk size is needed. The controller will translate this into enough
+    // pods to create buffer not exceeding the total size above
+    PerChunkResources *ResourceLimits
+    // Optional if more detailed limitations than target object ref on its own is not enough.
+    NodeSelector NodeSelector
+}
+
+type PodBufferCapacity struct {
+    ReplicaNum int
+    PodTemplate PodTemplateRef
+}
+
+type NodeAutoscalingBufferStatus struct {
+    PodBufferCapacity   PodBufferCapacity
+	PodTemplateGeneration int	
+    Conditions           []BufferCondition
+    LastFullyProvisioned *metav1.Time
+}
+```
+
+### What is the responsibility split between the NodeAutoscalingBuffers controller and autoscaler?
+
+The controller would own the logic of translating the buffer configuration into
+a pod spec that would define what spare space needs to be provisioned in the
+cluster to fulfill the requirements of the buffer. The autoscaler component
+would use the pod spec to provision the capacity.
+
+In case of the cluster autoscaler we would deploy the controller as an optional
+subprocess:
+
+![Figure](./images/buffers-component-structure.jpg)
+
+This design has two main advantages compared to embedding all the logic in the
+library that would be used in autoscaler code:
+
+*   Responsibility split: it splits the responsibility of understanding buffer
```

Comment:
For the second bit here, I'm not quite sure what "picked up" means -- even if the controller has reconciled the new pod definition into the status, that still doesn't mean that the autoscaler has seen the new pod. This problem exists whether you use API to store the state or store the state internally in memory (the library way that you are talking about above) so I don't really see this as a tangible benefit of one over the other